### PR TITLE
Add OAuth sign in

### PR DESCRIPTION
**Pull request by rogerhutchings**
Adds the `signInWithOAuthRedirect` method to the auth module, allowing in-browser authentication via OAuth. I've named the method to allow for a future popup implementation as well.

To do:
- [x] Write OAuth sign in method
- [ ] Write tests

cc @brian-c

### CODE REVIEW

**Comment by rogerhutchings** [comment]
@brian-c I think this is ready to go. I haven't added a test for the `completeSignInWithOAuth` as it's basically reusing previously tested code, and I'm not entirely sure how to get a token on its own without going through the browser; if you have any ideas, let me know!

File: lib/auth.js [hunk 1]
```diff
@@ -61,7 +61,10 @@ module.exports = new Model({
   },
 
   _handleNewBearerToken: function(request) {
-    var response = JSON.parse(request.text);
```
Comment (brian-c) [inline lib/auth.js]: We should pass a single type into this function consistently. Is it JSON sometimes and not others?

File: lib/auth.js [hunk 2]
```diff
@@ -231,6 +234,47 @@ module.exports = new Model({
     }.bind(this));
   },
 
+  signInWithOAuthRedirect: function (args, window) {
```
Comment (brian-c) [inline lib/auth.js]: Overriding `window` seems like a bad thing. I also think it's reasonable to expect this to run in a certain context (the browser).

File: lib/auth.js [hunk 3]
```diff
@@ -231,6 +234,47 @@ module.exports = new Model({
...
+    // Check to see if we're running in the browser or not
+    if (typeof window === 'undefined') {
```
Comment (brian-c) [inline lib/auth.js]: `window` is declared as an argument, so you can compare it to `undefined` directly.

File: lib/auth.js (hunk 2 above)
Comment (rogerhutchings) [inline lib/auth.js]: It does, but it then makes it harder to test - I couldn't find a decent way of mocking the window object that didn't require using a global.

File: lib/auth.js (hunk 1 above)
Comment (rogerhutchings) [inline lib/auth.js]: I think it originally expected the response to be passed directly from the http request - we could parse it first and then call this function, negating the need for this line.

File: lib/auth.js [hunk 4]
```diff
@@ -231,6 +234,47 @@ module.exports = new Model({
...
+          args.appId,
+          '&redirect_uri=',
+          args.redirectUri
+        ].join('');
+        window.location.href = url;
+      }
+    }.bind(this));
+  },
+
+  completeSignInWithOAuth: function (args) {
```
Comment (brian-c) [inline lib/auth.js]: I'm not super familiar with OAuth: when is this function called?

Comment (rogerhutchings) [inline lib/auth.js]: So - after you finish signing in on the Zoo login page, it then redirects you back to your app. But it needs to deserialize the response (which is supposed to be passed as a set of URL query parameters, but in our case it leaves off the leading `?`, yay). 

I couldn't think of a way of handling that which doesn't involve the app - so my current solution is to look for a route that contains those params, and call this function which completes the sign in process.

File: lib/auth.js [hunk 5]
```diff
@@ -231,6 +234,47 @@ module.exports = new Model({
...
+        var url = [
+          config.host,
+          '/oauth/authorize',
+          '?response_type=token',
+          '&client_id=',
+          args.appId,
+          '&redirect_uri=',
+          args.redirectUri
+        ].join('');
+        window.location.href = url;
```
Comment (brian-c) [inline lib/auth.js]: If we use `location.assign(url)` here we won't have to rebuild and compare strings in the test.

Comment (rogerhutchings) [inline lib/auth.js]: How would it work instead - pass in `location` and stub that method?

File: lib/auth.js (hunk 4 above)
Comment (brian-c) [inline lib/auth.js]: Where does that happen though? In the app? Does the app have to do anything with that value, or could it be handled in this module?

File: lib/auth.js (hunk 5 above)
Comment (brian-c) [inline lib/auth.js]: Yeah, I think it'd be a function that calls `t.pass`.

File: test/auth.js [hunk 6]
```diff
@@ -146,3 +148,35 @@ test('Disabling an account works', function(t) {
         });
     });
 });
+
+test('First step of signin with OAuth works', function(t) {
+  t.plan(2);
+  t.throws(auth.signInWithOAuthRedirect, 'Should not work outside the browser');
+
+  var GOOD_OAUTH_LOGIN_DETAILS = {
+    appId: '24ad5676d5d25c6aa850dc5d5f63ec8c03dbc7ae113b6442b8571fce6c5b974c',
```
Comment (brian-c) [inline test/auth.js]: The app ID is per-custom app, right? Is this a test app for staging?

File: test/auth.js [hunk 7]
```diff
@@ -146,3 +148,35 @@ test('Disabling an account works', function(t) {
...
+    redirectUri: 'http://localhost',
```
Comment (brian-c) [inline test/auth.js]: I couldn't get this to work in the browser. I also tried `location.href` and `"http://localhost:3735"`, but only got "The redirect uri included is not valid."

File: lib/auth.js (hunk 2 above)
Comment (brian-c) [inline lib/auth.js]: Ideally we'd be able to run browser-specific things in a browser, but I'm not sure that's worth investigating right now (I have no idea how we'd run a test across different windows and origins). In PFE we're using a `window-shim` lib to render on the server. Might be overkill here.

File: lib/auth.js (hunk 4 above)
Comment (rogerhutchings) [inline lib/auth.js]: Yeah, in the app - the bit that goes with it is zooniverse/zoo-react-starterify#37. All it does is parse the response, then redirect you to a given URL. I'm all for handling it inside this module, I'm not a big fan of tightly coupling this process to the app. 

Just did a quick bit of research - I could watch for the `[hashchange](https://html.spec.whatwg.org/multipage/indices.html#event-hashchange)` event and process that instead?

**Comment by rogerhutchings** [comment]
I'm going to close this branch - I've had another go, and it's producing a much cleaner solution. It does require a backend change to make the token refresh process a bit smoother, though.

**Pull request closed without merging**